Add statistics library

kubark42 opened the pull request:
This adds a statistics library to Tau Labs. It has several functions:
1. Calculate linear mean and standard deviation
2. Calculate linear variance
3. Calculate Pearson's coefficient
4. Calculate circular mean and standard deviation
5. Calculate angular deviation
6. Calculate a Bayesian update (concerning dynamic systems)

In addition, there is a full set of unit tests. They have been tested against MATLAB (with the caveat below that linear standard deviation has an efficiency shortcut[*] ).

I wrote this library for internal analysis of sensors and have been very happy with the in-flight results. I am submitting it in hopes that others find it useful.

@mlyle You might find this useful as well. I'm especially keen to know if you see interesting compiler optimizations that I've missed.
## 

[*] The linear standard deviation is calculated using the less frequent division by N instead of N-1. This is theoretically less accurate, but for the purposes of evaluating standard deviation over 100 samples only leads to a loss of 1% accuracy, which seems worthwhile if this function will be executed frequently.
##

kubark42 committed:
Statistics: Add unit test framework

***NOTE*** This runs no tests. The purpose of this commit is to lay the
groundwork for the unit test infrastructure.

kubark42 committed:
Statistics: Add linear statistics unit tests

Tests:
*) Succussfully initialize a structure with a full data array
*) Succussfully initialize a structure with a partial data array
*) Successfully increment a structure with a full data array
*) Successfully increment a structure with a partial data array
*) Calculate mean
*) Calculate standard deviation

kubark42 committed:
Statistics: Add circular statistics unit tests

Tests:
*) Succussfully initialize a structure with a full data array
*) Succussfully initialize a structure with a partial data array
*) Successfully increment a structure with a full data array
*) Successfully increment a structure with a partial data array
*) Calculate mean
*) Calculate standard deviation

kubark42 committed:
Statistics: Added large vector to unit testing

This added linear and circular unit testing for a large (100 element)
array.

kubark42 committed:
Statistics: Functions for efficiently calculating mean and std. dev.

This recognizes that the standard deviation for a data set can be
calculated by the number of elements, a rolling sum of elements, and a
rolling sum of elements squared. The decision is made to use the slightly
less correct std. deviation calculation which divdes by N, instead of N-1.
For large data sets, this makes only a small scalar difference. This
results in a slight gain of computational efficiency.

kubark42 committed:
Statistics: Add directional statistics

This adds support for calculating statistics on directional fields, such as
a circle.

kubark42 committed:
Statistics: Add Pearson Correlation

1) Adds a linear variance function, which is very similar to the standard
deviation function. The advantage is that this is cheaper to calculate than
the standard deviation.
2) Added a function to calculate the Pearson correlation, which can be
thought of as a normalized covariance between two variables.

kubark42 commented on flight/Libraries/math/statistics.c:
```diff
@@ -0,0 +1,352 @@
+float get_linear_variance(const struct linear_mean_and_std_dev *X)
+{
+	// sigma^2 = T2/T0 - (T1/T0)^2
+	float variance = (X->T2*X->T0 - X->T1*X->T1) / (X->T0*X->T0);
```
This can also be calculated as 

```
float mean_X = X->T1 / X->T0;
float variance = X->T2/X->T0 - mean_X*mean_X;
```

which saves two multiplications but adds a division. 

I'm unsure which is the generally more optimal approach. I'm open to arguments either way.

kubark42 commented:
@solidgoldbomb, if you have a moment, could you briefly review my unit test framework and tell me if I'm violating any best-practices for unit tests? I'm still a little hazy on where best to draw the line on avoiding code duplication and yet still fully testing the functions.

mlyle commented on flight/Libraries/math/statistics.c:
```diff
@@ -0,0 +1,352 @@
+	// or +Inf, depending on from which direction the first autovariance value is
+	// approaching 0. Rather than try to calculate this, we simply return an impossible
+	// value to indicate which
+	if (x_autovariance < 1e-5) {
```
f suffix on all of these

mlyle commented on flight/Libraries/math/statistics.c:
```diff
@@ -0,0 +1,352 @@
+
+	for (int i=0; i<num_samples; i++) {
+		// Sum of elements
+		X->T1 += (double)y[i];
```
The cast is implied and has no effect.  There's a lot of these through the existing Tau code and I suggest they be avoided in general (having a cast effectively neuters type checking in the future)-- not that it matters so much here.

kubark42 replied:
~~I think this was from back when I was experimenting with trying to make floats work. Thanks for catching it.~~

Leaving this in to be explicit, esp. in case the data format changes later.

mlyle commented on flight/Libraries/math/statistics.c:
```diff
@@ -0,0 +1,352 @@
+ * @param p_H1_Ux_H0 The probability of state H0 becoming H1 after the control is applied
+ * @param p_H1_Ux_H1 The probability of state H1 staying H1 after the control is applied
+ */
+void bayes_filter(double *belief_H0, double *belief_H1,
```
Some of the stuff below, like the circular mean filters, really needs precision and it's hard to envision running it on a large data set at single precision with larger values.  

This particular one seems to be overkill at double, and using it at double would put it out of reach for most targets.  On the other hand this type of Bayesian filter is routinely run at half-float precision on GPUs for hypothesis generation in scientific computing.  Would it be possible to move this one to float?

(All targets could benefit from an are-we-flying filter for various types of safety improvement, but relatively few could spare this implementation-- requiring either two implementations for different target sizes, neglecting the flash-constrained F3 and F1 targets, or this change).

Alternatively, one could #define a type before including this file, which in turn is typedef'd and used in the prototypes.

kubark42 replied:
> really needs precision and it's hard to envision running it on a large data set at single precision with larger values.

Yeah, for sure! In fact, in my experience all the statistics functions should go to pure doubles. Single precision causes headaches because things look like they should work and then wind up not in practice because of simple design decisions like increasing the number of sample points. Bayesian hypothesis testing was one of those areas, because even five 9's of certainty is low (unless you have an excellent understanding of your PDF, which is unlikely and not particularly valuable in many scenarios).

I think it'd be cool if someone were to realize there's a neat application on the F1 and then use this code as the basis for a reduced precision version. If there were a specific, well-bounded question that we were looking to ask I bet there's a good chance we could to optimize all the math to fit. But I wrote this to be a generic solution so it's a little early for optimization. The risk of reducing precision is that people will use the lower precision in a knee-jerk fashion, not realizing the ramifications, and wind up having a miserable experience.

mlyle replied:
Well, the example is something we definitely need for various things-- enhanced failsafe and "air mode" need to have a good estimate of whether we're in flight.  Which could be heuristic or could be based on hypothesis filtering.

And, I don't know.  For many things, 2-3 9's is great, and float is good for 6+.  Especially in many circumstances we'll have a great initial prior ("almost certainly not flying"), fairly high certainty observations along the way (high throttle command?  sustained non-unity acceleration?  varying altimeter?)  and thresholds using the test are likely to be somewhere like 90-95%.  Certainly it would be better to encourage the use of something like this-- even lacking some precision-- instead of other heuristics for efficiency.

kubark42 replied:
YMMV, but single-precision floats crapped out for me long before I had a reliable usage around five 9's. Not saying you can't get there, just that my tests had sufficient variance that important details went missing and I ultimately didn't feel I had much confidence in my "belief" due to the numerical imprecision. Things improved once I moved to doubles (much like it does with Kalman Filters).

I think the suitability for a single-precision version will have to be determined on a case-by-case scenario. At this early stage, it's probably safest to err to the side of "works in all cases, but isn't optimally efficient".

mlyle replied:
OK.  Just note that in the things I'm thinking we'll use this for-- in dRonin the lack of the ability to specify will either force us to diverge on this (editing it to let type be specified) or not use this library.

kubark42 replied:
Understood. Because it's currently unhooked to anything, I'm expecting you won't bring this over until you have a solid use for it, at which time you'll have resolved for that usage whether singles are acceptable.

The submission is to get people's brains thinking about all the ways statistics can be powerful, and give them a unit-tested library to try stuff out. I'm excited to see what people come up with. Glad you're taking a look.

mlyle commented on flight/Libraries/math/statistics.c:
```diff
@@ -0,0 +1,352 @@
+		X->T1 += (double)y[i];
+
+		// Sum of squares of elements
+		X->T2 += (double)y[i]*(double)y[i];
```
1.0 \* ... can be used for type promotion without the scariness of a cast on this line, though I've got less of an objection to this one.

kubark42 replied:
Yeah, this one is still necessary as long as the input value is float. It's arguable that the input values themselves don't need anywhere near the precision of a float, so long as the statistical analysis has double precision.

mlyle replied:
Yes, 1.0 \* ... will promote to double though without a cast (and without the downsides of casting).  OTOH it's slightly more arcane.

kubark42 replied:
~~Fair enough. I'll change it. It is probably less likely to be accidentally changed than a double anyway.~~

Fails to compile because of double promotion. Leaving it as is.

mluessi commented on flight/Libraries/math/statistics.c:
```diff
@@ -0,0 +1,352 @@
+	float std_dev;
+
+	// sigma = sqrt(T0*T2 - T1^2)/T0
+	double radicand = X->T0*X->T2 - X->T1*X->T1;
```
Shouldn't the first `T0` be `T2`?

kubark42 replied:
You mean `X->T2*X->T2 - X->T1*X->T1`? No, it needs to be `X->T0`. It's the consequence of some nice series multiplications.

kubark42 committed:
Statistics: Add Bayesian filter

This is a bayesian test which handles dynamic cases. Inspired by
Probablistic Robotics, Chap. 2.4, "Bayes Filters".

peabody124 commented:
I'll review this and should be able to merge it this weekend. 

Just out of curiosity, the copyright line is your name. Would you be willing to change it to TL? I'm not sure if we want to start adding all individuals in that section, ideally.

kubark42 commented:
> Just out of curiosity, the copyright line is your name. Would you be willing to change it to TL? I'm not sure if we want to start adding all individuals in that section, ideally.

Thanks for catching this. This is because until our code is released externally it's not under GPL. I can add the header in, but if we wanted to start down the path of #2155, this would be a great place to try it out. Thoughts?

peabody124 commented:
@kubark42 thats fine with me, although might slow down this merge

kubark42 commented:
@peabody124 I'm fine with that. This PR is pretty much guaranteed to be conflict free so I'm not worried. If that's the only remaining issue I can always go with the embedded license if time becomes critical.

peabody124 commented:
Want to fix your header and we can merge this?

kubark42 commented:
Yup. I'll do that as soon as the parent repo reemerges from the mandatory 10 day counter-notice period.

kubark42 committed:
Statistics: Relicense to GPLv3

kubark42 commented:
Frivolous DMCA take-down successfully refuted, so can push to repo again. Upshot: new commit relicensing code to GPLv3, merge when you're ready.
